bin/curriculum/backfill-allow-multiple-attempts.rb

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env ruby
+
+require_relative '../../deployment'
+raise unless [:development, :adhoc, :levelbuilder].include? rack_env
+
+# Wait until after initial error checking before loading the rails environment.
+def require_rails_env
+  puts "loading rails environment..."
+  start_time = Time.now
+  require_relative '../../dashboard/config/environment'
+  puts "rails environment loaded in #{(Time.now - start_time).to_i} seconds."
+end
+
+require_rails_env
+
+def level_is_standalone?(level)
+  level.script_levels.count > 0
+end
+
+def level_is_contained?(level)
+  level.levels_parent_levels.count {|l| l.kind == 'contained'} > 0
+end
+
+def find_union(standalone_levels, contained_levels)
+  standalone_level_names = standalone_levels.map(&:name)
+  contained_levels.filter {|l| standalone_level_names.include?(l.name)}.uniq
+end
+
+def backfill_free_response_levels
+  standalone_free_response_levels = FreeResponse.all.filter {|l| level_is_standalone?(l) && l.allow_multiple_attempts.nil?}
+  contained_free_response_levels = FreeResponse.all.filter {|l| level_is_contained?(l) && l.allow_multiple_attempts.nil?}
+  free_response_levels_to_flag = find_union(standalone_free_response_levels, contained_free_response_levels)
+  contained_free_response_levels -= free_response_levels_to_flag
+  puts "Free Response levels to manually check:"
+  free_response_levels_to_flag.each {|l| puts l.name}
+
+  contained_free_response_levels.each do |level|
+    level.allow_multiple_attempts = "false"
+    level.save!
+  end
+
+  FreeResponse.all.filter {|l| l.allow_multiple_attempts.nil?}.each do |level|
+    level.allow_multiple_attempts = "true"
+    level.save!
+  end
+end
+
+# Add `allow_multiple_attempts true/false` to the DSL text for the level
+# Then call assign_attributes to update the level and rewrite the file
+# This method catches exceptions as not all errors are immediately addressable
+def update_dsl_level(level, allow_multiple_attempts)
+  begin
+    allow_multiple_attempts_str = allow_multiple_attempts.to_s
+    text = level.dsl_text
+    if text.include?('allow_multiple_attempts')
+      return if text.include?("allow_multiple_attempts #{allow_multiple_attempts_str}")
+      text.gsub!(/allow_multiple_attempts.*$/, "allow_multiple_attempts #{allow_multiple_attempts_str}")
+    else
+      text += "\nallow_multiple_attempts #{allow_multiple_attempts_str}"
+    end
+    level.assign_attributes({dsl_text: text, allow_multiple_attempts: allow_multiple_attempts_str})
```

Comment:
I think I'd love to see a screenshot of what a dsl level looks like and how this text addition works, just for my own learnings!

Reply:
This is what the edit page of a DSL level looks like:

![Screen Shot 2023-04-03 at 9 16 19 AM](https://user-images.githubusercontent.com/46464143/229520937-a4b74fbe-42cc-425c-b632-58864daabbf6.png)

Essentially, any edits need to be made in that top box under "DSL Editor". If you edit the markdown under "Edit Markdown", it automatically updates the dsl text.

Then you press save (bottom right), which hits the `levels_controller`, which [calls](https://github.com/code-dot-org/code-dot-org/blob/ff60b588a6102d5cc09dc43258a82c1463cbda9c/dashboard/app/controllers/levels_controller.rb#L287) `assign_atttributes`. That in turn [calls](https://github.com/code-dot-org/code-dot-org/blob/ff60b588a6102d5cc09dc43258a82c1463cbda9c/dashboard/app/models/levels/dsl_defined.rb#L204) `create_from_level_builder`, which as far as I can tell, should really be named `create_or_update_from_level_builder`. Neither Match nor Multi levels override this function so the logic used in in [DSLDefined](https://github.com/code-dot-org/code-dot-org/blob/ff60b588a6102d5cc09dc43258a82c1463cbda9c/dashboard/app/models/levels/dsl_defined.rb#L109). That method both [updates or creates an entry in the database](https://github.com/code-dot-org/code-dot-org/blob/ff60b588a6102d5cc09dc43258a82c1463cbda9c/dashboard/app/models/levels/dsl_defined.rb#L92) and [writes to the serialized DSL file](https://github.com/code-dot-org/code-dot-org/blob/ff60b588a6102d5cc09dc43258a82c1463cbda9c/dashboard/app/models/levels/dsl_defined.rb#L133).